### PR TITLE
Update Google auth approach to use new libraries

### PR DESCRIPTION
**Pull request by daikema**
**NOTE: This is a breaking change for those using Google auth and will require updates to your config and Google auth registration as well as a frontend using the code associated with the [corresponding frontend pull request](https://github.com/CARTAvis/carta-frontend/pull/2345) to work.**

The Google auth approach now has the controller accept an ID token from Google at a callback address, and then based on this generates locally-issued refresh tokens + access tokens similar to the other supported authentication approaches as since [per the migration docs](https://developers.google.com/identity/oauth2/web/guides/migration-to-gis#session_state) we now need to maintain session state.

Use of this updated controller requires changes to both the Google oauth client registration and the local controller configuration.

Adding the callback URL to the [oauth client configuration in the Google API console is needed](https://console.developers.google.com/apis/credentials).  If modifying an existing client, you'll need to add a "Authorized redirect URI".  If your app is installed at `https://example.com`, this would be `https://example.com/api/auth/googleCallback`.

As far as the local configuration is concerned the controller new requires the existance of a public/private keypair to sign its tokens, which can be generated [as already described in the documentation](https://carta-controller.readthedocs.io/en/latest/configuration.html#authentication).

The `google` block in the `authProviders` section of your CARTA config should be updated to look something like the following:
```
        "google":
                 {
                     "clientId": "my-client-id.apps.googleusercontent.com",
                     "validDomain": "example.com",
                     "publicKeyLocation": "/etc/carta/carta_public.pem",
                     "privateKeyLocation": "/etc/carta/carta_private.pem",
                     "issuer": "carta.example.com",
                     "userLookupTable": "/etc/carta/userlookup.txt",
                 }
```
`clientId` should be the value shown in the Google dashboard, and `validDomain` can be used to used to restrict the list of permitted users to those associated with a particular Google workspace / cloud organizational account.  Note that user lookups retain a similar functionality to what was there before, defaulting to the use of mapping using the verified email addresses returned for the user.  (`publicKeyLocation`, `privateKeyLocation`, and `issuer` now appear similar to how they appear when using other mechanisms, and should be similar filled in using the locations for the public/private keypair and how CARTA will identify itself in the tokens it's signing, with the hostname as a suggestion).

### CODE REVIEW

**Comment by kswang1029** [comment]
@acdo2002 please help to deploy this to ASIAA’s instance for testing.

**Comment by daikema** [comment]
I probably add that for these tests to work, given that the frontend pull request has yet to be merged in, you'll need to  add a field like `"frontendPath": "/my/home/dir/carta-frontend/build"` to your CARTA `config.json` to get a corresponding frontend paired, where the directory in question contains a prebuilt frontend from the other pull request.

File: docs/src/introduction.rst [hunk 1]
```diff
@@ -25,11 +25,12 @@ Detailed installation instructions are available for :ref:`Ubuntu<focal_instruct
 Authentication support
 ----------------------
 
-The CARTA controller supports four modes for authentication. All four modes use refresh and access tokens, as described in the `OAuth2 Authorization flow <https://tools.ietf.org/html/rfc6749#section-1.3.1>`_, stored in `JWT <https://jwt.io/>`_ format. The modes are:
+The CARTA controller supports five modes for authentication. All four modes use refresh and access tokens, as described in the `OAuth2 Authorization flow <https://tools.ietf.org/html/rfc6749#section-1.3.1>`_, stored in `JWT <https://jwt.io/>`_ format. The modes are:
```
Comment (veggiesaurus) [inline docs/src/introduction.rst]: need to change the second `four` as well 😇

File: src/auth/google.ts [hunk 2]
```diff
@@ -1,36 +1,82 @@
...
 
-        // Use either the email or the unique sub ID as the username
+        // Do the mapping
         const username = authConf.useEmailAsId ? payload?.email : payload?.sub;
 
         // check that username exists and email is verified
         if (!username || !payload?.email_verified) {
             console.log("Google auth rejected due to lack of unique ID or email verification");
-            return undefined;
+            return res.status(500).json({'error': 'An error occured processing your login'});
```
Comment (veggiesaurus) [inline src/auth/google.ts]: seem to be a few inconsistent uses of `'` versus `"` here

**Review by veggiesaurus — CHANGES_REQUESTED** [review]
just minor changes 🚀

File: docs/src/introduction.rst (hunk 1 above)
Comment (daikema) [reply docs/src/introduction.rst]: Have to see what other documentation issues others pick up.  Could probably stand a more thorough overhaul.

File: src/auth/google.ts (hunk 2 above)
Comment (daikema) [reply src/auth/google.ts]: Standardized on `"`

**Comment by daikema** [comment]
Should probably add that with the move to always handle logout using redirection, this code is now a breaking change for all auth methods (not just Google) and should only be used in conjunction with [the corresponding frontend pull request](https://github.com/CARTAvis/carta-frontend/pull/2345).

**Review by veggiesaurus — APPROVED** [review]

**Review by acdo2002 — APPROVED** [review]
ASIAA set up this new carta-controller PR (+ carta-frontend PR) and it works fine.

**Pull request merged**